services/pddb/src/main.rs

```diff
@@ -413,6 +415,21 @@ struct TokenRecord {
     pub alloc_hint: Option<usize>,
     pub conn: Option<xous::CID>, // callback connection, if one was specified
 }
+
+struct FileHandle {
+    pub dict: String,
+    pub key: String,
+    pub basis: Option<String>,
+    pub alloc_hint: Option<usize>,
+    pub offset: u64,
+    pub length: u64,
+    pub conn: Option<xous::CID>, // callback connection, if one was specified
+
+    /// This is set to `true` when a file is removed in order to prevent
+    /// other operations from functioning.
+    pub deleted: bool,
```

Comment:
The `deleted` is a proxy for when you have a file open in multiple handles. Really, we should look through every open handle in the system rather than just for this process.

When a file is deleted, all operations on that file should fail. This is in contrast to something like Linux where the file is only garbage-collected when all handles close. It's debatable which approach is better, but for a PDDB it seems preferrable for files to go away immediately.

I wasn't sure what the callback was used for, and unfortunately there are now two places where callback data is stored. And they're not compatible. I'd like to unify these two approaches somehow.

Reply:
Ooh...yes, the PDDB would assume that the view is consistent across processes. At the moment we don't do much cross-process stuff with files but I could see a future where, for example, the `wlan` SSIDs are managed by one process but read by another.

I think all handles failing immediately when a handle is deleted is the correct approach.

The callback is still more experimental in my code anyways, but, yes, the idea is that if you have say, a secret Basis open, and then you close it, you want a way to notify the application program that "you should not display that handle anymore". A concrete example might be that you have a list of TOTP authenticators running, one of them exists in a secret basis. With the basis open, you'd see that secret TOTP authenticator in the `vault` app. But if you close that secret basis, that authenticator entry /should/ disappear in the `vault` app.

Like I said, at the moment, it's just a very coarse, big hammer where the callback triggers a message to be sent to the main loop that forces a reload of all the data from the disk. i think in practice a per-file object handle (as it currently is made out to be) is too fine-grained and doesn't even really make sense. In other words, some refactoring to come. Issue #194 vaguely tracks a problem I've also observed, it's more just a reminder that I need to write more test cases and flesh it out. I'm happy to hear more opinions on how to do this better.